docs/tutorial/algorithm_interfaces.rst

```diff
@@ -1,2 +1,464 @@
+:meth:`~.AlgorithmInterface.train_step`                 :math:`[B,\ldots]` or :math:`[TB,\ldots]`    ``inputs``, ``state``, ``rollout_info``  ``output``, ``state``, ``info`` Off-policy
+:meth:`~.AlgorithmInterface.calc_loss`                  :math:`[T,B,\ldots]`                         ``info``                                 ``loss_info``                   On/Off-policy
+:meth:`~.AlgorithmInterface.after_update`               :math:`[T,B,\ldots]`                         ``root_inputs``, ``info``                ``None``                        On/Off-policy
+:meth:`~.AlgorithmInterface.after_train_iter`           :math:`[T,B,\ldots]`                         ``root_inputs``, ``rollout_info``        ``None``                        On/Off-policy
+======================================================  ===========================================  =======================================  =============================== ===============
+
+:math:`T` is :attr:`.TrainerConfig.unroll_length` for on-policy algorithms and
+:attr:`.TrainerConfig.mini_batch_length` for off-policy algorithms. :math:`B` is
+:attr:`.create_environment.num_parallel_environments` for on-policy algorithms and
+:attr:`.TrainerConfig.mini_batch_size` for off-policy algorithms. We can see that
+the first three interfaces are called at every time step, while the last three
+are called once for multiple time steps. The off-policy pipeline has one
+additional :meth:`~.AlgorithmInterface.train_step` interface compared to on-policy.
+
+.. note::
+
+    If :attr:`.TrainerConfig.temporally_independent_train_step=True`, we will call
+    :meth:`~.AlgorithmInterface.train_step` on steps along a temporal extent in
+    parallel. So the tensor shape is :math:`[TB,\ldots]`. This speeds up the
+    training forward step and the result will be reshaped to
+    :math:`[T,B,\ldots]` when fed to :meth:`~.AlgorithmInterface.calc_loss`.
+
+In most cases, we only need to override these interfaces while following their
+input/output signatures. Potentially we are able to accomplish many things through
+them.
+
+Argument conventions
+--------------------
+
+Here we assume that there is only a single algorithm in the pipeline. In fact, multiple
+algorithms could consist of an *algorithm hierarchy* so that the calling of
+their interfaces of the same name is performed in a nested way. We leave this
+advantaged usage to a later chapter :doc:`./customize_algorithms`.
+
+* ``inputs`` denotes the observations/states based on which
+  an algorithm's interface does computation at every step.
+* ``output`` is specially used to represent a step function's main output, e.g.,
+  action of an RL algorithm, latent embedding of an encoder algorithm, or goal
+  representation of a goal proposal algorithm.
+* ``state`` denotes episodic memory (e.g., RNN state) that is automatically reset
```

Comment:
episodic memory usually refers to long term memory (e.g. replay buffer). `state` for algorithm is more like working memory.

Reply:
changed to working memory